libsql_client/dbapi2/types.py

```diff
@@ -0,0 +1,1423 @@
+        return tokens[0].upper() in cls._dml_statements
+
+    @property
+    def is_readonly(self) -> bool:
+        # mimics sqlite3_stmt_readonly()
+        return not self.is_dml
+
+
+def check_thread(fn: Callable[P, T]) -> Callable[P, T]:
+    """Decorator that checks :py:class:`Connection` thread usage.
+
+    Otherwise raises ``ProgrammingError``.
+
+    :meta private:
+    """
+    def wrapper(*args: P.args, **kwargs: P.kwargs) -> T:
+        self = args[0]
+        assert len(args) > 0 and isinstance(self, Connection)
+        self._do_check_thread()
+        return fn(*args, **kwargs)
+
+    return wrapper
+
+
+def check_connection(fn: Callable[P, T]) -> Callable[P, T]:
+    """Decorator that checks :py:class:`Connection` being valid.
+
+        Otherwise raises ``ProgrammingError``.
+
+    :meta private:
+    """
+    def wrapper(*args: P.args, **kwargs: P.kwargs) -> T:
+        self = args[0]
+        assert len(args) > 0 and isinstance(self, Connection)
+        self._do_check_connection()
+        return fn(*args, **kwargs)
+
+    return wrapper
+
+
+class Connection(metaclass=ABCMeta):
+    """Implement :py:class:`sqlite3.Connection` for remote servers.
+
+    :meta private:
+    """
+    _closed: bool
+    _thread: Optional[int]
+    _database: str
+    _timeout: float
+    _detect_types: int
+    _isolation_level: Optional[IsolationLevel]
+    _check_same_thread: bool
+    _cached_statement_getter: Callable[[str], Statement]
+    _in_transaction: bool
+    _cursors: WeakSet
+    _autocommit: Autocommit
+    _trace_callback: Optional[TraceCallback]
+    _limits: Dict[int, int]
+    row_factory: Optional[RowFactory]
+    text_factory: Optional[TextFactory]
+    cursor_factory: CursorFactory
+
+    def __init__(
+        self,
+        database: str,
+        timeout: float,
+        detect_types: int,
+        isolation_level: Optional[IsolationLevel],
+        check_same_thread: bool,
+        cached_statements: int,
+        autocommit: Autocommit,
+    ) -> None:
+        check_valid_autocommit(autocommit)  # required by test
+        self._closed = False
+        self._thread = threading.current_thread().ident
+        self._database = database
+        self._timeout = timeout
+        self._detect_types = detect_types
+        self._check_same_thread = check_same_thread
+        self._cached_statement_getter = functools.lru_cache(
+            cached_statements
+        )(self.__call__)
+        self._in_transaction = False
+        self._cursors = WeakSet()
+        self.row_factory = None
+        self.text_factory = None
+        self.cursor_factory = Cursor
+        self._autocommit = autocommit
+        self._trace_callback = None
+        self._limits = _default_limits.copy()
+        self.isolation_level = isolation_level
+        self._raw_init()  # after this point we can call the DB
+        if autocommit is False:
+            self.cursor()._query_execute("BEGIN", want_rows=False)
```

Comment:
it will skip some of the autocommit logic, it impacts when you do a commit/rollback, executescript.

It's following exactly what they do, so you may refer to their C python bindings

Reply:
Let's hope that people won't use this mode in practice, it won't work well with Turso :)

Reply:
their test suite passes 🤷